### PR TITLE
Kalin nonchev protein dl grch38

### PR DESCRIPTION
**Pull request by KalinNonchev**

### CODE REVIEW

**Comment by brechtmann** [comment]
@MuhammedHasan please review and merge this branch

File: kipoiseq/extractors/protein.py [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+    return Interval(str(row.Chromosome),
+                    int(row.Start),
+                    int(row.End),
+                    strand=str(row.Strand),
+                    attrs={'tag': str(row.tag)})
+
+
+class CDSFetcher:
+
+    def __init__(self, gtf_file, vcf_matcher_df=None):
```
[the remark this answers is not in the document]
Comment (KalinNonchev) [reply kipoiseq/extractors/protein.py]: these are the  transcript_ids, which have variants. If this class is called from ProteinVCFSeqExtractor we do not need all transcript_ids in the gtf file, but only those which have variants.

[the remark this answers is not in the document]
Comment (MuhammedHasan) [reply kipoiseq/extractors/protein.py]: You need to explain this to user who do not have any ideas about the codebase.

[the remark this answers is not in the document]
Comment (MuhammedHasan) [reply kipoiseq/extractors/protein.py]: This is not sure passing vcf_matcher_df object the best way. But I cannot imagine a better way of doing it.

[the remark this answers is not in the document]
Comment (MuhammedHasan) [reply kipoiseq/extractors/protein.py]: Let's discuss this point in code review.

File: kipoiseq/extractors/protein.py [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+
+    def extract_all(self):
+        """
+        Extract all amino acid sequences for transcript_ids with variants
+        given into the vcf_file
+        """
+        for transcript_id in self.cds_fetcher.transcripts:
+            yield self.extract(transcript_id)
+
+    def extract_list(self, list_with_transcript_id: 'list of str'):
```
[the remark this answers is not in the document]
Comment (KalinNonchev) [reply kipoiseq/extractors/protein.py]: it is implemented, but it breaks the ci tests

**Comment by MuhammedHasan** [comment]
```python
def read_gtf(f, full=True, as_df=False, nrows=None, duplicate_attr=False):

df = pyranges.read_gtf(gtf_file, output_df=True, duplicate_attr=True)
```

replace `output_df`  argument name with `as_df`

File: kipoiseq/extractors/protein.py [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+        print("Ambiguous start and end! Skip seq!")
+        seq = "NNN"  # NNN will be translated as empty string
+    elif "cds_end_NF" not in tag and "cds_start_NF" not in tag and len(seq) % 3 != 0:
+        print("No tags for ambiguous start and end, but len % 3 != 0. Skip seq!")
+        seq = "NNN"  # NNN will be translated as empty string
+
+    return seq
+
+
+def gtf_row2interval(row):
```
Comment (MuhammedHasan) [inline kipoiseq/extractors/protein.py]:
```
def gtf_row2interval(row, interval_attrs: List[str] = None):
    """
    Convert gtf row object into interval class.
    """
    attrs = {i: row[i] for i in interval_attrs if i in row}
    return Interval(str(row.Chromosome),
                    int(row.Start),
                    int(row.End),
                    strand=str(row.Strand),
                    attrs=attrs)
gtf_row2interval(row, ['tag'])
```

**Pull request merged**